// @sanity/runtime-cli/dist/actions/blueprints/stacks.js
import config from '../../config.js';
import getHeaders from '../../utils/get-headers.js';
const { blueprints } = config.server;
export const stacksUrl = `${blueprints}vX/blueprints/stacks`;
export async function listStacks(auth) {
    const response = await fetch(stacksUrl, {
        method: 'GET',
        headers: getHeaders(auth),
    });
    const stacks = await response.json();
    return {
        ok: response.ok,
        error: response.ok ? null : stacks.error?.message,
        stacks,
    };
}
/** @deprecated Use getStack instead */
export async function getStackByName({ name, auth, }) {
    const { ok, stacks, error } = await listStacks(auth);
    if (!ok || !stacks) {
        return {
            ok: false,
            error: error || 'Failed to retrieve stacks',
            stack: null,
            stackId: null,
        };
    }
    const foundStack = stacks.find((stack) => stack.name === name);
    if (!foundStack) {
        return {
            ok: true,
            error: null,
            stack: null,
            stackId: null,
            availableStacks: stacks.map((s) => s.name),
        };
    }
    const stackResult = await getStack({ stackId: foundStack.id, auth });
    return {
        ok: stackResult.ok,
        error: stackResult.error,
        stack: stackResult.stack,
        stackId: foundStack.id,
    };
}
export async function getStack({ stackId, auth, }) {
    const response = await fetch(`${stacksUrl}/${stackId}`, {
        method: 'GET',
        headers: getHeaders(auth),
    });
    const stack = await response.json();
    return {
        ok: response.ok,
        error: response.ok ? null : stack.message,
        stack,
    };
}
export async function createStack({ stackPayload, auth, }) {
    const response = await fetch(stacksUrl, {
        method: 'POST',
        headers: getHeaders(auth),
        body: JSON.stringify(stackPayload),
    });
    const stack = await response.json();
    return {
        ok: response.ok,
        error: response.ok ? null : stack.message,
        stack,
    };
}
export async function updateStack({ stackId, stackPayload, auth, }) {
    const response = await fetch(`${stacksUrl}/${stackId}`, {
        method: 'PUT',
        headers: getHeaders(auth),
        body: JSON.stringify(stackPayload),
    });
    const stack = await response.json();
    return {
        ok: response.ok,
        error: response.ok ? null : stack.message,
        stack,
    };
}
export async function destroyStack({ stackId, auth, }) {
    const response = await fetch(`${stacksUrl}/${stackId}`, {
        method: 'DELETE',
        headers: getHeaders(auth),
    });
    const stack = await response.json();
    return {
        ok: response.ok,
        error: response.ok ? null : stack.message,
        stack,
    };
}
